File: hash_10xBD.py
print("\n running hash")
import os
os.environ['OPENBLAS_NUM_THREADS'] = '1'
import sys
import numpy as np
import minocore
from scipy.io import mmread

args = sys.argv
print("args=",args)
#args = ["dontuse", "/users/ndyjack/Dist_Proj/tables/test_datasets/", "/users/ndyjack/Dist_Proj/tables/hashes/", "1000", "1305694", "1306127"]
#mtx_dir = '/fastscratch/myscratch/ndyjack/tables/tmp_10xBD.4.mtx'
mtx_dir = args[2] + 'tmp_10xBD.' + args[4] + '.mtx'
output_dir = args[3] + 'tmp_10xBD.' + args[4] + '.csv'
seed = 1337

dat = mmread(mtx_dir) #read sparse matrix
dat = dat.toarray().transpose() #transpose and reformat
dat = dat / np.sum(dat, axis=1)[:,np.newaxis] #normalize
n_features = dat.shape[1] # get number of features

#hasher=args[5]
hasher='jsd'

if hasher == "cl1":
  hashf = minocore.ClippedL1LSHasher(dim=n_features, k=10, l=10, w=0.01,seed=seed) #set up hash function
  hashout = hashf.project(dat) #project data
elif hasher == "jsd":
  hashf = minocore.JSDLSHasher(dim=n_features, k=10, l=10, r=0.01,seed=seed) #set up hash function
  hashout = hashf.project(dat) #project data
elif hasher == "l1":
  hashf = minocore.L1LSHasher(dim=n_features, k=10, l=10, w=0.01,seed=seed) #set up hash function
  hashout = hashf.project(dat) #project data
elif hasher == "l2":
  hashf = minocore.L2LSHasher(dim=n_features, k=10, l=10, w=0.01,seed=seed) #set up hash function
  hashout = hashf.project(dat) #project data
elif hasher == "s2jsd":
  hashf = minocore.S2JSDLSHasher(dim=n_features, k=10, l=10, w=0.01,seed=seed) #set up hash function
  hashout = hashf.project(dat) #project data
else:
  sys.exit("invalid hash function")

hashout_round = np.floor(hashout).astype(np.uint16)

tmp = np.sum(hashout_round,1)
p = [i for i, e in enumerate(tmp) if e == 0]

#[i for i, e in tmp if e == 0]
# [item for item in tmp if item == 0]
#np.sum(hashout_round,1) == 0


np.savetxt(output_dir, hashout_round, delimiter=",",fmt='%i')#store hash as csv to do computations

#for x in np.sum(hashout_round,1):
#  print(x)

#delete the temporary mtx file

File: hash_10xBD.R
args <- commandArgs(trailingOnly=T)
cat("\n","running Rscript","\n")
cat("args=",args)
library(Matrix)
library(DelayedArray)
library(HDF5Array)

#args= c("/users/ndyjack/Dist_Proj/tables/test_datasets/", "/fastscratch/myscratch/ndyjack/tables/", "/fastscratch/myscratch/ndyjack/hashes/", "400", "521494", "522800NULL")
args[6] <- sub("NULL","",args[6])

#example args
#data dir, mtx dir, hash dir, i, istart, i end 
# cells = 1306127
# genes = 24015

#setRealizationBackend("RleArray")
setRealizationBackend("HDF5Array")
dat <- HDF5Array(paste0(args[1],"1M_neurons_full_expr.h5"),name = "10xBD_Full")
dat <- as.matrix(realize(dat[,as.integer(args[5]):as.integer(args[6])]))
dat <- Matrix(dat,sparse=T)
writeMM(dat,paste0(args[2],"tmp_10xBD.",args[4],".mtx"))

#dat <- h5mread(paste0(args[1],"1M_neurons_full_expr.h5"),name="10xBD_Full")
#block_dim <- c(24015,as.integer(args[5]) - as.integer(args[4]))
#view_port <- ArrayViewport(as.integer(c(24015,1306127)),IRanges(c(1,as.integer(args[4])),width=block_dim))
#tickmarks=list(1:1000,as.integer(args[4]):as.integer(args[5])))
#ArbitraryArrayGrid(tickmarks=list(1:1000,as.integer(args[4]):as.integer(args[5])))
#dat <- read_block(x=paste0(args[1],"1M_neurons_full_expr.h5"),viewport = view_port)
#dat <- as.matrix(counts(dat[,as.integer(args[4]):as.integer(args[5])]))
#dat <- Matrix(dat,sparse=T)
#writeMM(dat,paste0(args[1],"tmp_10xBD.",args[3],".mtx"))
